oshmem/mca/spml/ucx/spml_ucx.c

```diff
@@ -539,82 +570,83 @@ int mca_spml_ucx_ctx_create(long options, shmem_ctx_t *ctx)
-    if (NULL == ctx_item->ctx.ucp_peers) {
+    ucx_ctx->ucp_peers = (ucp_peer_t *) calloc(nprocs, sizeof(*(ucx_ctx->ucp_peers)));
+    if (NULL == ucx_ctx->ucp_peers) {
         goto error;
     }
 
+    if (mca_spml_ucx.active_array.ctxs_count == 0) {
+        opal_progress_register(spml_ucx_ctx_progress);
+    }
+
     for (i = 0; i < nprocs; i++) {
         ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
         ep_params.address    = (ucp_address_t *)(mca_spml_ucx.remote_addrs_tbl[i]);
-        err = ucp_ep_create(ctx_item->ctx.ucp_worker, &ep_params,
-                            &ctx_item->ctx.ucp_peers[i].ucp_conn);
+        err = ucp_ep_create(ucx_ctx->ucp_worker, &ep_params,
+                            &ucx_ctx->ucp_peers[i].ucp_conn);
         if (UCS_OK != err) {
             SPML_ERROR("ucp_ep_create(proc=%d/%d) failed: %s", i, nprocs,
                        ucs_status_string(err));
             goto error2;
         }
+
+        for (j = 0; j < MCA_MEMHEAP_SEG_COUNT; j++) {
+            mkey = &memheap_map->mem_segs[j].mkeys_cache[i][0];
+            ucx_mkey = &ucx_ctx->ucp_peers[i].mkeys[j].key;
+            err = ucp_ep_rkey_unpack(ucx_ctx->ucp_peers[i].ucp_conn,
+                                     mkey->u.data,
+                                     &ucx_mkey->rkey);
+            if (UCS_OK != err) {
+                SPML_UCX_ERROR("failed to unpack rkey");
+                goto error2;
+            }
+            mca_spml_ucx_cache_mkey(ucx_ctx, mkey, j, i);
+        }
     }
 
     SHMEM_MUTEX_LOCK(mca_spml_ucx.internal_mutex);
-
-    opal_list_append(&(mca_spml_ucx.ctx_list), &ctx_item->super);
-
+    _ctx_add(&mca_spml_ucx.active_array, ucx_ctx);
     SHMEM_MUTEX_UNLOCK(mca_spml_ucx.internal_mutex);
 
-    (*ctx) = (shmem_ctx_t)(&ctx_item->ctx);
-
+    (*ctx) = (shmem_ctx_t)ucx_ctx;
     return OSHMEM_SUCCESS;
 
  error2:
     for (i = 0; i < nprocs; i++) {
-        if (ctx_item->ctx.ucp_peers[i].ucp_conn) {
-            ucp_ep_destroy(ctx_item->ctx.ucp_peers[i].ucp_conn);
+        if (ucx_ctx->ucp_peers[i].ucp_conn) {
+            ucp_ep_destroy(ucx_ctx->ucp_peers[i].ucp_conn);
         }
     }
 
-    if (ctx_item->ctx.ucp_peers)
-        free(ctx_item->ctx.ucp_peers);
+    if (ucx_ctx->ucp_peers)
+        free(ucx_ctx->ucp_peers);
 
  error:
-    ucp_worker_destroy(ctx_item->ctx.ucp_worker);
-    OBJ_RELEASE(ctx_item);
+    ucp_worker_destroy(ucx_ctx->ucp_worker);
+    free(ucx_ctx);
     rc = OSHMEM_ERR_OUT_OF_RESOURCE;
     SPML_ERROR("ctx create FAILED rc=%d", rc);
     return rc;
 }
 
 void mca_spml_ucx_ctx_destroy(shmem_ctx_t ctx)
 {
-    mca_spml_ucx_ctx_list_item_t *ctx_item, *next;
-    size_t i, nprocs = oshmem_num_procs();
-
     MCA_SPML_CALL(quiet(ctx));
 
     SHMEM_MUTEX_LOCK(mca_spml_ucx.internal_mutex);
+    _ctx_remove(&mca_spml_ucx.active_array, (mca_spml_ucx_ctx_t *)ctx);
+    _ctx_add(&mca_spml_ucx.idle_array, (mca_spml_ucx_ctx_t *)ctx);
```

Comment:
you are removing context from progress, all outstanding operations should be flushed

Reply:
Does ucp_worker_flush depend on remote progress?
Since context destroy is a local operation, I'm guess we should call ucp_worker_flush_nb(), and check status when we actually destroy contexts in finalize?

Reply:
as I can see context is not destroyed here - it is just moved into idle array and all outstanding ops may be left uncompleted. ucp_worker_flush may depend from remote side for fetch_and_op calls, but as I can see user can't remove OSHMEM context when there are incomplete ops, like called ***_nbi and after this call ctx_destroy without call shmem_ctx_quiet... but I'm not sure about this :)

Reply:
it should be ok, because disconnect_nb (invoked by ctx_cleanup) will flush eps

Reply:
from the other side, flushing worker may help to avoid 'hard-to-debug' issues. I tend to agree with @hoopoepg that worker flush is worth to  be done here. @yosefe, wdyt?

Reply:
added flush() - @xinzhao3 need to run this through unit tests again